File: content/exercises/intermediate-python/pandas/groupby/checker.py
import io
import runpy
import sys


def main() -> None:
    submission_path = sys.argv[1]
    output_path = sys.argv[2]
    tests: list[tuple[str, bool, str]] = []
    feedback: list[str] = []
    status = "failed"

    try:
        namespace = runpy.run_path(submission_path, run_name="__main__")
        tests.append(("code runs", True, "Code executed without errors."))
    except Exception as error:  # noqa: BLE001
        namespace = {}
        tests.append(("code runs", False, "Code did not run successfully."))
        feedback.append(f"Execution error: {error}")

    if namespace:
        agg = namespace.get("agg")
        if agg is None:
            tests.append(("agg exists", False, "Define aggregation DataFrame as variable agg."))
        else:
            try:
                import numpy as np
                import pandas as pd

                is_df = isinstance(agg, pd.DataFrame)
                tests.append(("agg type", is_df, "agg should be a pandas DataFrame."))
                if is_df:
                    has_cols = list(agg.columns) == ["sum", "mean"]
                    tests.append(("agg columns", has_cols, "agg columns should be ['sum', 'mean']."))
                    expected_sum = {"Americas": 1754, "Asia": 4136, "EU": 2147}
                    expected_mean = {"Americas": 1754.0, "Asia": 2068.0, "EU": 1073.5}
                    index_ok = set(agg.index.tolist()) == set(expected_sum.keys())
                    tests.append(("agg index", index_ok, "agg should group by region."))
                    if index_ok and has_cols:
                        sums_ok = all(
                            float(agg.loc[key, "sum"]) == float(value)
                            for key, value in expected_sum.items()
                        )
                        means_ok = all(
                            np.isclose(float(agg.loc[key, "mean"]), float(value))
                            for key, value in expected_mean.items()
                        )
                        tests.append(("sum values", sums_ok, "sum values per region are incorrect."))
                        tests.append(("mean values", means_ok, "mean values per region are incorrect."))
            except Exception as error:  # noqa: BLE001
                tests.append(("pandas validation", False, f"Could not validate agg: {error}"))

    passed_all = tests and all(item[1] for item in tests)
    if passed_all:
        status = "passed"
        feedback.append("Correct solution submitted.")
    elif not feedback:
        feedback.append("Group by region and aggregate exports_bn with sum and mean.")

    output_stream = io.StringIO()
    output_stream.write(f"STATUS:{status}\n")
    for item in feedback:
        output_stream.write(f"FEEDBACK:{item}\n")
    for name, passed, message in tests:
        output_stream.write(f"TEST:{name}|{'pass' if passed else 'fail'}|{message}\n")

    with open(output_path, "w", encoding="utf-8") as handle:
        handle.write(output_stream.getvalue())


if __name__ == "__main__":
    main()

File: content/exercises/intermediate-r/dplyr/group-summarise/checker.R
args <- commandArgs(trailingOnly = TRUE)
if (length(args) < 2) stop("Usage: checker.R <submission_path> <output_path>")
submission_path <- args[[1]]
output_path <- args[[2]]

test_results <- list()
feedback <- c()
status <- "failed"
record_test <- function(name, passed, message) {
  test_results[[length(test_results) + 1]] <<- list(name = name, passed = passed, message = message)
}

env <- new.env(parent = baseenv())
execution_ok <- TRUE
execution_error <- NULL
last_value <- NULL
tryCatch({
  sourced <- source(submission_path, local = env)
  last_value <- sourced$value
}, error = function(e) {
  execution_ok <<- FALSE
  execution_error <<- conditionMessage(e)
})

if (!execution_ok) {
  record_test("code runs", FALSE, "Your code did not run successfully.")
  feedback <- c(feedback, paste("Execution error:", execution_error))
} else {
  record_test("code runs", TRUE, "Code executed without errors.")
  is_df <- is.data.frame(last_value)
  record_test("returns table", is_df, "Pipeline should return a summary table.")
  if (is_df) {
    has_cols <- all(c("cyl", "mean_mpg", "mean_hp") %in% colnames(last_value))
    cyl_sorted <- identical(last_value$cyl, sort(last_value$cyl))
    record_test("summary columns", has_cols, "Include cyl, mean_mpg, and mean_hp columns.")
    record_test("arranged by cyl", cyl_sorted, "Arrange results by cyl.")
  }
}

passed_all <- length(test_results) > 0 && all(vapply(test_results, function(t) isTRUE(t$passed), logical(1)))
if (passed_all) {
  status <- "passed"; feedback <- c(feedback, "Correct solution submitted.")
} else if (length(feedback) == 0) {
  feedback <- c(feedback, "Group by cyl, summarise mean_mpg/mean_hp, and arrange(cyl).")
}

output_lines <- c(paste0("STATUS:", status))
for (msg in feedback) output_lines <- c(output_lines, paste0("FEEDBACK:", msg))
for (t in test_results) output_lines <- c(output_lines, paste0("TEST:", t$name, "|", if (isTRUE(t$passed)) "pass" else "fail", "|", t$message))
writeLines(output_lines, con = output_path)
